### DISCUSSION TITLE
Help needed: how to integrate observable framework with SSG / SSR for pre-rendering thousands of framework dashboards?

### DISCUSSION THREAD

**Discussion by jaanaltosaar** [observablehq/framework#855] [Q&A] [answered]
What would be the easiest way to integrate the observable framework with vite's server side rendering / static site generation based on dynamic routes known at build time? 

For example, I am using Vike for pre-rendering HTML pages on build for dynamic routes for every public use microdata area in the United States: https://vike.dev/pre-rendering 

Use case:

* I generate thousands of dashboards (dependent on thousands of <1 MB parquet files on Amazon S3). 
* The schema are all known ahead of time - ETL transforms version-controlled in `dbt` [here](https://github.com/jaanli/exploring_american_community_survey_data/) if needed. (This is for an open source project to make data the American Community Survey from the Census Bureau more accessible.)
* There are at least 4974 possible dashboards  so far, with likely more on the way. They are generated using jinja2 and Mosaic, which is [supported](https://github.com/observablehq/framework/discussions/834) by observable framework. 
* Example dashboard prototype: https://s13.gifyu.com/images/SCGH2.gif (code [here](https://github.com/jaanli/lonboard/blob/example-american-community-survey/examples/american-community-survey.ipynb) with data loaded from parquet files.)

Happy to collaborate / chat if others are interested in this, and happy to contribute to the project if this is not supported yet.

P.S. Loving the framework development experience so far :) this is like Christmas came early!

**Comment by Fil** [comment, marked as answer]
You can reference thousands of files in a project, by listing all of them as static references:

````md
```js
const files = [
  FileAttachment("data1.csv"),
  FileAttachment("data2.csv"),
// …
  FileAttachment("data1000.csv")
];
```
````

(use a script to build this list); it will build the files into the site.

And you can then use (download) those files “dynamically” with FileAttachment (as csv, text, or as a DuckDB source), for example with an input to select the *id* of the data source:

````md
```js
const FA = FileAttachment;
const id = view(Inputs.range([1, 1000], { step: 1 }));
```

```js
display(FA(`data${id}.csv`));
const data = FA(`data${id}.csv`).csv({ typed: true });
```

```js
display(Inputs.table(data));
```

```js
const db = DuckDBClient.of({ name: FA(`data${id}.csv`).csv({ typed: true }) });
```

```js
display(await db.sql`SELECT * FROM name`);
```
````

Another way to set a dynamic id it through the page's URL: page?id=12
````md
```js
const id = +(new URL(document.location).searchParams.get("id") ?? "1"); // 12
```
````



Note: with the current implementation of FileAttachment, the name of the file is predictable, so you could use `fetch("/_file/data"+id+".csv")`; however this is not guaranteed, in fact we plan to introduce content-hash filenames soon (https://github.com/observablehq/framework/issues/260), meaning that the only way for the client to know the URL of a file will be to use the FileAttachment helper.


HTH.

(I'm discovering `vike` with your message, so I can't help with that. But your project looks 👍)

**Reply by jaanaltosaar** [reply]
You're the best, thanks so much @Fil - excited to try this out :D and will ping you when ready!

**Comment by davidgasquez** [comment]
I think I'm hitting a similar issue.

I have a list of IDs I need to build a dashboard for. E.g: `mydash.com/user/1`, `mydash.com/user/2`, `mydash.com/user/3`... Right now, I was able to cram all the users data into a 600MB parquet file and serve a website covering them all using the page URL like `mydash.com/user?id=1`. 

Wondering if there is a way to (1) skip having to upload and read a 600MB file (only reads 150MB at the end) with each user and (2) also have the `mydash.com/user/1` prerendered at build time so the browser only has to load a page with static tiles/charts.

Any ideas?

**Reply by mbostock** [reply]
Please upvote https://github.com/observablehq/framework/issues/245 if you’re interested in this feature.

One way to solve this problem now is to write a script that generates Markdown pages as part of your build process. We do this in the documentation to generate a page for each theme. For example the [coffee theme preview](https://observablehq.com/framework/theme/coffee) is generated by the [`generate-themes.ts`](https://github.com/observablehq/framework/blob/main/docs/theme/generate-themes.ts) script which runs immediately prior to the build command:

https://github.com/observablehq/framework/blob/41f874c93bab2bce9db9b373a7144d2bf7ca5835/package.json#L23

**Reply by davidgasquez** [reply]
Thanks @mbostock! Will check that approach out. That said, it would be awesome if this (parametrized pages rendered at buildtime) could be done at build time natively to avoid querying the large parquet file each time the page opens.

**Reply by jaanaltosaar** [reply]
> Thanks @mbostock! Will check that approach out. That said, it would be awesome if this (parametrized pages rendered at buildtime) could be done at build time natively to avoid querying the large parquet file each time the page opens. 
> 
> 
> 
> 
> 
> 
> 
> 

+1! @onefact, we have been using @vercel as a workaround with the help of their team:

https://github.com/onefact/incremental-static-regeneration-demo

Demo here. It queries an AWS Lambda function that is kept alive on a lambda and whose only job is to return the results of `JOIN`s or other long-running queries. 

Happy to help if others are also blocked on this and need guidance - we will be open sourcing our work here soon, and open to collaboration/contribution/sponsorship.

**Reply by davidgasquez** [reply]
Is there any ETA on the SSG plot rendering? 

I'm wondering if it makes sense for my project to go and use something like Astro and render things there using Observable Plot (tried something like that [here](https://davidgasquez.github.io/astro-static-dashboard/)) or I should wait for these features to ship to Observable Framework.

Ideally, I'd love for Framework to render the ~8k pages (surfacing basic metrics and charts without filters) at build time and embed the minimum amount of JSON in the OJS charts and inputs so that opening `mydash.com/user/1` doesn't load DuckDB WASM and has then to query a big parquet file. 

Right now, following the [`generate-themes.ts`](https://github.com/observablehq/framework/blob/main/docs/theme/generate-themes.ts) script approach gives me a bunch of pages that still need to request DuckDB WASM and query the large parquet file.

**Reply by jaanaltosaar** [reply]
+1 great idea @davidgasquez ! We need the same. We've prototyped with NextJS: https://github.com/onefact/incremental-static-regeneration-demo using an always-on Lambda to retrieve the queries faster than WASM. Could a hybrid solution like that work? 

Happy to chat with your team next week as it seems we are doing the same thing.

**Reply by jaanaltosaar** [reply]
Cc @sumanthratna for visibility

**Comment by mbostock** [comment]
Framework 1.11.0 adds [parameterized routes](https://observablehq.com/framework/params) and [page loaders](https://observablehq.com/framework/page-loaders) which we recommend for this use case.
